### PR TITLE
IGNITE-16619 IndexQuery should support limit

### PR DESCRIPTION
**Pull request by yurinaryshkin**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+    /** */
+    @Parameterized.Parameter(5)
+    public String idxName;
+
+    /** Number of duplicates of indexed value. */
+    @Parameterized.Parameter(6)
+    public int duplicates;
+
+    /** */
+    @Parameterized.Parameters(name = "qryPar={0} atomicity={1} mode={2} node={3} backups={4} idxName={5} duplicates={6}")
```
Comment (timoninmaxim) [inline modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java:80]: Too many params for this simple functionality. Let's make test simple, like tests for other params - IndexQueryLocalTest, IndexQueryKeepBinaryTest.

It's enough to check single cache, single index, single backup, with/without duplicates.

File: modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryTestSuite.java [hunk 2]
```diff
@@ -46,7 +46,8 @@
     ThinClientIndexQueryTest.class,
```
Comment (timoninmaxim) [inline modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryTestSuite.java:46]: Let's test the functionality with thin client too.

File: modules/core/src/main/java/org/apache/ignite/cache/query/IndexQuery.java [hunk 3]
```diff
@@ -152,6 +178,27 @@ public String getIndexName() {
...
+    }
+
+    /**
+     * Sets limit to response records count.
+     *
+     * @param limit If 0 or less, considered to be no limit.
+     * @return {@code this} For chaining.
+     */
+    public IndexQuery<K, V> setLimit(int limit) {
+        this.limit = limit;
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/cache/query/IndexQuery.java:197]: Let's ensure that limit is positive value.

**Review by timoninmaxim — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/cache/query/IndexQuery.java [hunk 4]
```diff
@@ -94,11 +97,34 @@ public IndexQuery(Class<?> valCls, @Nullable String idxName) {
...
+    }
+
+    /**
+     * Specify index with cache value class, index name and limit.
+     *
+     * @param valCls Cache value class.
+     * @param idxName Index name.
+     * @param limit Limits response records count. If 0 or less, considered to be no limit.
+     */
+    public IndexQuery(Class<?> valCls, @Nullable String idxName, int limit) {
```
[the remark this answers is not in the document]
Comment (yurinaryshkin) [reply modules/core/src/main/java/org/apache/ignite/cache/query/IndexQuery.java:110]: removed the constructors

File: modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java (hunk 1 above)
Comment (yurinaryshkin) [reply modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java:80]: simplified

File: modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryTestSuite.java (hunk 2 above)
Comment (yurinaryshkin) [reply modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryTestSuite.java:46]: while adding test discovered that thin client didn't pass the limit parameter, resolved that

File: modules/core/src/main/java/org/apache/ignite/cache/query/IndexQuery.java [hunk 5]
```diff
@@ -152,6 +155,30 @@ public String getIndexName() {
...
+    }
+
+    /**
+     * Sets limit to response records count.
+     *
+     * @param limit POsitive limit to set.
+     * @return {@code this} For chaining.
+     */
+    public IndexQuery<K, V> setLimit(int limit) {
+        if (limit <= 0)
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/cache/query/IndexQuery.java:174]: Please, use `A.ensure` for such checks

File: modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+        IndexQuery<Long, Person> qry = new IndexQuery<>(Person.class, IDX);
+
+        qry.setLimit(limit);
+
+        assertEquals(limit, qry.getLimit());
+    }
+
+    /** */
+    private void checkRangeQueries(int duplicates) throws Exception {
+
```
Comment (timoninmaxim) [inline modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java:123]: Please remove this empty line.

**Review by timoninmaxim — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCacheIndexQueryRequest.java [hunk 7]
```diff
@@ -57,6 +57,8 @@ public ClientCacheIndexQueryRequest(BinaryRawReaderEx reader) {
 
         int part = reader.readInt();
 
+        int limit = reader.readInt();
```
[the remark this answers is not in the document]
Comment (yurinaryshkin) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCacheIndexQueryRequest.java]: Added new feature INDEX_QUERY_LIMIT and corresponding checks on client and server

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCacheIndexQueryRequest.java [hunk 8]
```diff
@@ -47,8 +49,10 @@ public class ClientCacheIndexQueryRequest extends ClientCacheQueryRequest {
 
     /**
      * @param reader Reader.
+     * @param protocolCtx
      */
-    public ClientCacheIndexQueryRequest(BinaryRawReaderEx reader) {
+    public ClientCacheIndexQueryRequest(BinaryRawReaderEx reader,
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCacheIndexQueryRequest.java:54]:
```suggestion
    public ClientCacheIndexQueryRequest(
        BinaryRawReaderEx reader,
        ClientProtocolContext protocolCtx
    ) {
```

File: modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        CacheConfiguration<Long, Person> ccfg = new CacheConfiguration<Long, Person>()
+                .setName(CACHE)
+                .setIndexedTypes(Long.class, Person.class)
+                .setAtomicityMode(TRANSACTIONAL)
+                .setCacheMode(REPLICATED)
+                .setQueryParallelism(1)
+                .setBackups(0);
```
Comment (timoninmaxim) [inline modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java:77]: No need to set, backups=0 is default value.

File: modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        CacheConfiguration<Long, Person> ccfg = new CacheConfiguration<Long, Person>()
+                .setName(CACHE)
+                .setIndexedTypes(Long.class, Person.class)
+                .setAtomicityMode(TRANSACTIONAL)
+                .setCacheMode(REPLICATED)
+                .setQueryParallelism(1)
```
Comment (timoninmaxim) [inline modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java:76]: No need to set, 1 is the default value

File: modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+
+        assertEquals(limit, qry.getLimit());
+    }
+
+    /** */
+    private void checkRangeQueries(int duplicates) throws Exception {
+        // Query empty cache.
+        IndexQuery<Long, Person> qry = new IndexQuery<>(Person.class, IDX);
+
+        assertTrue(crd.cache(CACHE).query(qry).getAll().isEmpty());
```
Comment (timoninmaxim) [inline modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java:119]: Let's skip this check, it doesn't test the limit functionality.

File: modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+
+            checkLimit(criterion, limit, left, right, duplicates);
+        }
+    }
+
+    /** */
+    private void checkLimit(IndexQueryCriterion criterion, int limit, int left, int right, int duplicates) throws Exception {
+        IndexQuery<Long, Person> qry = new IndexQuery<>(Person.class, IDX);
+
+        if (criterion != null) {
```
Comment (timoninmaxim) [inline modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java:157]: Please remove curly braces here.

File: modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+            qry.setCriteria(criterion);
+        }
+        qry.setLimit(limit);
+
+        QueryCursor<Cache.Entry<Long, Person>> cursor = crd.cache(CACHE).query(qry);
+
+        int expSize = (right - left) * duplicates;
+
+        if (limit > 0 && limit < expSize) {
```
Comment (timoninmaxim) [inline modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java:166]: And here

File: modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+                expKeys.add((long)CNT * j + i);
+                if (expOrderedValues.size() >= limit)
+                    break loop;
+            }
+        }
+
+        AtomicInteger actSize = new AtomicInteger();
+
+        ((QueryCursorEx<Cache.Entry<Long, Person>>)cursor).getAll(entry -> {
+
```
[the remark this answers is not in the document]
Comment (timoninmaxim) [reply modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java:187]: Empty line is still here

File: modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+        if (limit > 0 && limit < expSize) {
+            expSize = limit;
+        }
+
+        Set<Long> expKeys = new HashSet<>(expSize);
+        List<Integer> expOrderedValues = new LinkedList<>();
+
+        IntUnaryOperator op = (i) -> i + 1;
+
+        loop: for (int i = left; i != right; i = op.applyAsInt(i)) {
```
Comment (timoninmaxim) [inline modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java:175]: `op.applyAsInt(i)` can be replaced with `i++`

File: modules/core/src/main/java/org/apache/ignite/internal/client/thin/TcpClientCache.java [hunk 16]
```diff
@@ -1016,6 +1016,15 @@ private QueryCursor<Cache.Entry<K, V>> indexQuery(IndexQuery<K, V> qry) {
                 w.writeBoolean(qry.isLocal());
                 w.writeInt(qry.getPartition() == null ? -1 : qry.getPartition());
 
+                if (!payloadCh.clientChannel().protocolCtx().isFeatureSupported(ProtocolBitmaskFeature.INDEX_QUERY_LIMIT)) {
+                    if (qry.getLimit() > 0) {
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/client/thin/TcpClientCache.java:1020]: Please remove curly braces here

File: modules/core/src/main/java/org/apache/ignite/internal/client/thin/TcpClientCache.java [hunk 17]
```diff
@@ -1016,6 +1016,15 @@ private QueryCursor<Cache.Entry<K, V>> indexQuery(IndexQuery<K, V> qry) {
...
+                        throw new ClientFeatureNotSupportedByServerException(ProtocolBitmaskFeature.INDEX_QUERY_LIMIT);
+                    }
+                }
+                else {
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/client/thin/TcpClientCache.java:1024]: And here

**Review by timoninmaxim — CHANGES_REQUESTED** [review]

File: modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (yurinaryshkin) [reply modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java]: removed

File: modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+        // All
+        checkLimit(null, 0, CNT, duplicates);
+
+        String fld = "id";
+
+        int pivot = new Random().nextInt(CNT);
+
+        // Lt.
+        checkLimit(lt(fld, pivot), 0, pivot, duplicates);
+
```
Comment (timoninmaxim) [inline modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java:127]: Please remove empty line

File: modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+    private void checkRangeQueries(int duplicates) throws Exception {
+        IndexQuery<Long, Person> qry = new IndexQuery<>(Person.class, IDX);
+
+        // Add data
+        insertData(duplicates);
+
+        // All
+        checkLimit(null, 0, CNT, duplicates);
+
+        String fld = "id";
```
Comment (timoninmaxim) [inline modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java:121]: Can be inlined

File: modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+
+    /** */
+    @Test
+    public void testRangeQueriesWithDuplicates() throws Exception {
+        checkRangeQueries(10);
+    }
+
+    /** */
+    @Test
+    public void testSetLimit() throws Exception {
```
Comment (timoninmaxim) [inline modules/indexing/src/test/java/org/apache/ignite/cache/query/IndexQueryLimitTest.java:95]: Exception doesn't throw

**Review by timoninmaxim — APPROVED** [review]

**Pull request merged**